### PR TITLE
Define workers for production apicast

### PR DESCRIPTION
**Pull request by eguzki**
https://issues.redhat.com/browse/THREESCALE-6449
                                            
                                   
* Production apicast workers defined in the CR. Another option would be to add it in the CM. 
* Staging apicast does not have option to tune workers. As this parameter is meant for performance tuning, it should not be necessary.

example
```
apiVersion: apps.3scale.net/v1alpha1                                    
kind: APIManager                                                        
metadata:                                                               
  name: apimanager1                                                     
spec:                                                                   
  wildcardDomain: example.com                     
  apicast:                                                              
    productionSpec:                                                     
      workers: 4
```

### CODE REVIEW

**Comment by miguelsorianod** [comment]
The option has been added but if someone modifies it then no changes happen at all. It behaves like a deployment-only attribute.
Is that intentional? should we reconcile it? Same kind of scenario will happen with the other request we've had about the log level

**Comment by eguzki** [comment]
Good point. I was just testing reconciliation and noticed that. I assumed env vars where reconciled (the apicast operator does, hence my mistake).

Should we reconcile specific env vars or the entire array? Maybe we can try reconciling the entire array but with a merge approach, which would allow adding custom env var and the operator would not revert it.

File: pkg/3scale/amp/operator/apicast_reconciler.go [hunk 1]
```diff
@@ -125,6 +135,25 @@ func (r *ApicastReconciler) Reconcile() (reconcile.Result, error) {
 	return reconcile.Result{}, nil
 }
 
+func (r *ApicastReconciler) apicastProductionWorkersEnvVarMutator(desired, existing *appsv1.DeploymentConfig) bool {
```
Comment (eguzki) [inline pkg/3scale/amp/operator/apicast_reconciler.go:138]: Not sure this is the best way to do this. wdyt @miguelsorianod ?

File: doc/apimanager-reference.md [hunk 2]
```diff
@@ -102,6 +102,7 @@ Generated using [github-markdown-toc](https://github.com/ekalinin/github-markdow
 | Affinity | `affinity` | [v1.Affinity](https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#affinity-v1-core) | No | `nil` | Affinity is a group of affinity scheduling rules |
 | Tolerations | `tolerations` | \[\][v1.Tolerations](https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#toleration-v1-core) | No | `nil` | Tolerations allow pods to schedule onto nodes with matching taints |
 | Resources | `resources` | [v1.ResourceRequirements](https://v1-17.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.17/#resourcerequirements-v1-core) | No | `nil` | Resources describes the compute resource requirements. Takes precedence over `spec.resourceRequirementsEnabled` with replace behavior |
+| Workers | `workers` | integer | No | auto | Defines the number of worker processes |
```
Comment (miguelsorianod) [inline doc/apimanager-reference.md:105]: If `auto` it's a possible value then Workers is not really an integer.
The way for users to set it as `auto` would then be telling them "leave this empty if you want auto be used". In that case, would be great to clarify in the doc

File: pkg/reconcilers/deploymentconfig.go [hunk 3]
```diff
@@ -13,71 +13,42 @@ import (
...
 )
 
-func DeploymentConfigResourcesMutator(existingObj, desiredObj common.KubernetesObject) (bool, error) {
-	existing, ok := existingObj.(*appsv1.DeploymentConfig)
-	if !ok {
-		return false, fmt.Errorf("%T is not a *appsv1.DeploymentConfig", existingObj)
+// DCMutateFn is a function which mutates the existing DeploymentConfig into it's desired state.
+type DCMutateFn func(desired, existing *appsv1.DeploymentConfig) bool
+
+func DeploymentConfigMutator(opts ...DCMutateFn) MutateFn {
```
Comment (miguelsorianod) [inline pkg/reconcilers/deploymentconfig.go:19]: I like this approach :+1:, makes easier adding different mutators in a single mutator

File: pkg/reconcilers/deploymentconfig.go [hunk 4]
```diff
@@ -140,3 +108,30 @@ func DeploymentConfigContainerResourcesReconciler(desired, existing *appsv1.Depl
 
 	return update
 }
+
+func DeploymentConfigEnvVarMergeMutator(desired, existing *appsv1.DeploymentConfig) bool {
```
Comment (miguelsorianod) [inline pkg/reconcilers/deploymentconfig.go:112]: Not sure about this having the https://github.com/3scale/3scale-operator/pull/548 where it only reconciles the log envvar (https://github.com/3scale/3scale-operator/pull/557).
If we are not sure probably would be a good idea to be safe and just reconcile the worker for now. We can always change to reconcile everything later

File: pkg/3scale/amp/operator/apicast_reconciler.go (hunk 1 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/operator/apicast_reconciler.go:139]: See https://github.com/3scale/3scale-operator/pull/548/files#r553952791 comment

**Review by miguelsorianod — CHANGES_REQUESTED** [review]
:+1: . I left some comments

File: doc/apimanager-reference.md (hunk 2 above)
Comment (eguzki) [reply doc/apimanager-reference.md]: The value in the `Default value` field has the implicit meaning that the value will be set when the field is empty/unset. No need to state it explicitly. 

Updated "auto" for more meaningful content with link to apicast upstream doc.

File: pkg/reconcilers/deploymentconfig.go (hunk 4 above)
Comment (eguzki) [reply pkg/reconcilers/deploymentconfig.go:112]: removing this method for another one focused in a single env var.

File: pkg/3scale/amp/operator/apicast_reconciler.go (hunk 1 above)
Comment (eguzki) [reply pkg/3scale/amp/operator/apicast_reconciler.go:137]: Updating for a new implementation based on reconciliation of a single env var

**Comment by eguzki** [comment]
Answered all comments and applied changes, ready for a new review @miguelsorianod

File: doc/apimanager-reference.md (hunk 2 above)
Comment (miguelsorianod) [reply doc/apimanager-reference.md]: :+1:

File: pkg/reconcilers/deploymentconfig.go (hunk 4 above)
Comment (miguelsorianod) [reply pkg/reconcilers/deploymentconfig.go]: :+1:

File: pkg/3scale/amp/operator/apicast_reconciler.go (hunk 1 above)
Comment (miguelsorianod) [reply pkg/3scale/amp/operator/apicast_reconciler.go:137]: :+1:

**Review by miguelsorianod — APPROVED** [review]

**Pull request merged**